Changed rule: valid-jsdoc — Set preferred tags where there are synonyms

niedzielski opened the pull request:
Set preferred tags for all synonyms listed[0]. The Closure Compiler tags
are also supported by default but no aliases were noted[1].

All changes were copied from the Services' Node.js config[2] except
yield, which was added.

http://usejsdoc.org/#block-tags
https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#jsdoc-tags
https://github.com/wikimedia/eslint-config-node-services/blob/master/.eslintrc.js#L219-L238

Krinkle commented on .eslintrc.json:
```diff
@@ -97,7 +97,34 @@
+      "prefer": {
+        "abstract": "virtual",
+        "augments": "extends",
+        "constructor": "class",
```
I don't think these are the same?

It's been useful to differentiate in documentation between the class (as a whole) and the specific method signature of the constructor. We do this for PHP code with Doxygen, and currently for (the majority of) our code with JSDuck, in which `@class` and `@constructor` are different concepts. If this relates to JSDoc, then we should postpone this until later. I would also expect JSDoc to be able to make the same distinction, though perhaps through different means.

Example:

* <https://doc.wikimedia.org/mediawiki-core/master/js/#!/api/mw.Uri>
* <https://doc.wikimedia.org/mediawiki-core/master/js/#!/api/mw.Title>

niedzielski replied:
Maybe we should use `@constructs` instead?

http://usejsdoc.org/tags-class.html
http://usejsdoc.org/tags-constructs.html

Krinkle replied:
In the present, that would produce an error with JSDuck.

Our current subset of tags are documented at <https://www.mediawiki.org/wiki/Manual:Coding_conventions/JavaScript#Tags>, with some projects defining additional custom tags and/or using JSDoc instead.

Once <https://phabricator.wikimedia.org/T187672> is further along,  we can revisit the defaults, but not yet.

We could, however, make a jsdoc-specific mixin, similar to the one we just landed for QUnit.

You'd use it like this:

```
{
  "extends:" [
    "wikimedia",
    "wikimedia/jsdoc"
  ]
}
```

Or something like that.

Meanwhile, we can also then add a stricter set of rules (rather than just a safe default subset) for JSDuck, under `wikimedia/jsduck`.

niedzielski replied:
If this is the only issue, how about I just omit `"constructor": "class",` for the time being and it can be added later when JSDuck is replaced ([T138401](https://phabricator.wikimedia.org/T138401))? If you're ok with that, I'll update these tags according to the docs you've linked to.

Thanks for sharing your note on wikimedia/jsduck. I would like to see better JSDuck usage of course but I personally would prefer to invest my own energy converting repos to JSDoc instead.

Krinkle replied:
Sounds good!

niedzielski committed:
test: Fix ESLint directive regex

`new RegExp('//|/*').test('/')` incorrectly evaluates to true since the
asterisk is unescaped.

niedzielski commented on test/fixtures/invalid-es5.js:
```diff
@@ -244,9 +244,33 @@ var APP;
 	};
+
+	/**
+	 * {@linkcode APP}
```
These two inline tags don't seem to be linted properly at this time.

Krinkle commented on test/fixtures/valid-es5.js:
```diff
@@ -234,10 +234,37 @@
+	 * @desc Description.
+	 * @external String
+	 * @file Description.
+	 * @emits Event
```
According to Codesearch, there are no uses of `@emits` in WMF-deployed repo. Instead, we use `@fires`.

niedzielski replied:
:+1: done.

Krinkle committed:
Changed rule: `valid-jsdoc` – Add various preferred tags

Set preferred tags for all synonyms from <http://usejsdoc.org/#block-tags>.

The Closure Compiler tags are also supported by default but no aliases
were noted. Per
<https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#jsdoc-tags>.

All changes were copied from the Services' Node.js config except
yield, which was added.
<https://github.com/wikimedia/eslint-config-node-services/blob/c48908abb7/.eslintrc.js#L219-L238>.

Limit preferred JSDoc tags to be JSDuck compatible

The following are now preferred:

- @virtual -> @abstract
- @var -> @member
- @prop -> @property

The following are open ended for now:

- @class / @constructor / @constructs
- @method / @func (and @func is preferred over @function)

Closes #80.
